cli/azd/pkg/azsdk/zip_deploy_client.go

```diff
@@ -114,6 +122,202 @@ func (c *ZipDeployClient) BeginDeploy(
 	return runtime.NewPoller(response, c.pipeline, pollerOptions)
 }
 
+// Deploys the specified application zip to the azure app service using deployment status api and waits for completion
+func (c *ZipDeployClient) BeginDeployTrackStatus(
+	ctx context.Context,
+	zipFile io.ReadSeeker,
+	subscriptionId,
+	resourceGroup,
+	appName string,
+) (*runtime.Poller[armappservice.WebAppsClientGetProductionSiteDeploymentStatusResponse], error) {
+	request, err := c.createDeployRequest(ctx, zipFile)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := c.pipeline.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	if !runtime.HasStatusCode(response, http.StatusAccepted) {
+		return nil, runtime.NewResponseError(response)
+	}
+
+	client, err := armappservice.NewWebAppsClient(subscriptionId, c.cred, nil)
+
+	if err != nil {
+		return nil, fmt.Errorf("creating web app client: %w", err)
+	}
+
+	deploymentStatusId := response.Header.Get("Scm-Deployment-Id")
+	if deploymentStatusId == "" {
+		return nil, fmt.Errorf("empty deployment status id")
+	}
+
+	// nolint:lll
+	// Example definition: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/web/resource-manager/Microsoft.Web/stable/2022-03-01/examples/GetSiteDeploymentStatus.json
+	var poller *runtime.Poller[armappservice.WebAppsClientGetProductionSiteDeploymentStatusResponse]
+	err = retry.Do(
+		ctx,
+		retry.WithMaxRetries(3, retry.NewConstant(5*time.Second)),
+		func(ctx context.Context) error {
```

Comment:
I think this [comment](https://github.com/Azure/azure-dev/pull/3965#discussion_r1625033368) may not be fully addressed yet, we're still using the `retry` library here:

> It may look a little cleaner if we use the retry options for `azure-sdk-for-go`. I believe the corresponding change would look like the following for your use case here. Note that we use `policy.WithRetryOptions` to create a `context.Context` that only affects the current request, and not the overall client.
> 
> ```go
> 	retryCtx := policy.WithRetryOptions(ctx, policy.RetryOptions{
> 		MaxRetries: 4,
> 		StatusCodes: append([]int{
> 			http.StatusRequestTimeout,      // 408
> 			http.StatusTooManyRequests,     // 429
> 			http.StatusInternalServerError, // 500
> 			http.StatusBadGateway,          // 502
> 			http.StatusServiceUnavailable,  // 503
> 			http.StatusGatewayTimeout,      // 504
> 		}, http.StatusNotFound), // 404
> 	})
> 	// nolint:lll
> 	// Example definition: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/web/resource-manager/Microsoft.Web/stable/2022-03-01/examples/GetSiteDeploymentStatus.json
> 	poller, err := client.BeginGetProductionSiteDeploymentStatus(retryCtx, resourceGroup, appName, deploymentStatusId, nil)
> 	if err != nil {
> 		return nil, fmt.Errorf("getting deployment status: %w", err)
> 	}
> ```

Reply:
The main reason why I would choose azure-sdk-for-go's built-in retry is to avoid complexity (what I meant by "it may look a little cleaner"). We have a retry mechanism built in into the SDK, and we should opt for using it instead of something else if possible. 

When we use different retry mechanisms, we need to guarantee that we do not create unexpected retry loops due to retries on top of retries, which has been a source of bugs in many software. Stacking retries easily changes the behavior the retry delay to exponential.